From patchwork Tue Jul 28 09:17:59 2026
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Chen-Yu Tsai <wenst@chromium.org>
X-Patchwork-Id: 14715041
Received: from mail-pl1-f182.google.com (mail-pl1-f182.google.com
 [209.85.214.182])
	(using TLSv1.2 with cipher ECDHE-RSA-AES128-GCM-SHA256 (128/128 bits))
	(No client certificate requested)
	by smtp.subspace.kernel.org (Postfix) with ESMTPS id 2A38941E6BA
	for <linux-clk@vger.kernel.org>; Tue, 28 Jul 2026 09:18:21 +0000 (UTC)
Authentication-Results: smtp.subspace.kernel.org;
 arc=none smtp.client-ip=209.85.214.182
ARC-Seal: i=1; a=rsa-sha256; d=subspace.kernel.org; s=arc-20240116;
	t=1785230303; cv=none;
 b=I34gm9pOzEQqK8sutsM50JyjKx6/Ka+InB7ppv66RuF3Hq0NH/TxFUyQy3WvvdgImULYseMP7b5P3Er8TJ9XVrDc8CYmvenVnx6qQyKSezHXvKnvVe87bMpPZt4gj4+VnMnRYrpfN8VUN8uQRxf0CtUbnXKaA7vA3GZZz0ubCCo=
ARC-Message-Signature: i=1; a=rsa-sha256; d=subspace.kernel.org;
	s=arc-20240116; t=1785230303; c=relaxed/simple;
	bh=oWT0z3sYbdBIxj39xHfrckXBnQbxT1C0dn7AE/zBdBM=;
	h=From:To:Cc:Subject:Date:Message-ID:In-Reply-To:References:
	 MIME-Version;
 b=Md/R4Skf6tGm3kDnbhB+B3HFXn1kzXwtzg82T5Cm7MECflW9pHD0ybJ/2uoLairOqerxaQgcjnjlIEyCkeN76+8XCrv38+p0fu1kCleGs26lrA4CfPZinMOkzZZJOMqNkAMZZGkbmL81t+it0/hBCRTyuLoS/AUihSKyOa1+Nyg=
ARC-Authentication-Results: i=1; smtp.subspace.kernel.org;
 dmarc=pass (p=none dis=none) header.from=chromium.org;
 spf=pass smtp.mailfrom=chromium.org;
 dkim=pass (1024-bit key) header.d=chromium.org header.i=@chromium.org
 header.b=A+M4A/+d; arc=none smtp.client-ip=209.85.214.182
Authentication-Results: smtp.subspace.kernel.org;
 dmarc=pass (p=none dis=none) header.from=chromium.org
Authentication-Results: smtp.subspace.kernel.org;
 spf=pass smtp.mailfrom=chromium.org
Authentication-Results: smtp.subspace.kernel.org;
	dkim=pass (1024-bit key) header.d=chromium.org header.i=@chromium.org
 header.b="A+M4A/+d"
Received: by mail-pl1-f182.google.com with SMTP id
 d9443c01a7336-2cc7ef7ec27so45796585ad.1
        for <linux-clk@vger.kernel.org>; Tue, 28 Jul 2026 02:18:21 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=chromium.org; s=google; t=1785230301; x=1785835101;
 darn=vger.kernel.org;
        h=content-transfer-encoding:mime-version:references:in-reply-to
         :message-id:date:subject:cc:to:from:from:to:cc:subject:date
         :message-id:reply-to:content-type;
        bh=0T75/EutYIQThS/ZfnwY3OlWCqIaGeaNzqSXOb58pGA=;
        b=A+M4A/+daG/izkUbu3m/UXYOPiz6PxVrLZ5jKcdoLN2/ixgDVIJcKrAFksznPWetvg
         qLLKk004p0TH/t1yIZr/CaHX3Hlx7kqDJ1YXWf3ARgMzITS0DRdIl5SBjy9dkrwYGKc9
         ka041sFlfnnP3OZJGq0Nav97KMKuKtiFubWfE=
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20251104; t=1785230301; x=1785835101;
        h=content-transfer-encoding:mime-version:references:in-reply-to
         :message-id:date:subject:cc:to:from:x-gm-gg:x-gm-message-state:from
         :to:cc:subject:date:message-id:reply-to:content-type;
        bh=0T75/EutYIQThS/ZfnwY3OlWCqIaGeaNzqSXOb58pGA=;
        b=YNRqTAwb0fhZwhgB7K4C/ZZd5jo6Xgp6eCl7Mwkn3G5sp/YOAZmsJMz94JMlg7r5yh
         oe4tMpJoviwdR2oFkpnIYGbIWudSHt3jPSny9IKk8wwHatmwaKrerSwnSN5cgTeCmo83
         zYgkh6+5sSxp8XvvhsJksiKC/vwpOaPWFgpj6fqoKXIVXH/+ykyT8wpmT2Z5sXv/6FFI
         38WUjzD8VoFV3ZnFroU2mMz7StlzI7bF9ACCWPD0LIMj5a7SW2suiqh1CzBvdu5a6Dni
         JUDIltoJMrhr1Hs/Sbal7e/n/DoP+U8spXQr5YUd42KCVgohXWaoz0/DeAqc9ksZ0ZeT
         lkSQ==
X-Forwarded-Encrypted: i=1;
 AHgh+Rrng9ewg0pu8hRRTFNHflKTIN3TNhemTQc92GRLSatOiOR7ikUnnIujETSiPb2TEWc6iQUlSgD8eOo=@vger.kernel.org
X-Gm-Message-State: AOJu0Yyim5v7iKeZgu80GTHuwIEph1qEjWNn0hmTDCz8OSfYmtF/jxV5
	K+1O0SDoOuV7y29R/CEjzhxp+TRva4t2lQnmhOchqdLc106NWVmc7FoPm0uIP7cD4Q==
X-Gm-Gg: AR+sD11bePZ8a9gpM0FJ4zI3tCYqKdTqKgCz8nFO9wTkoC00LR+ZEiqLI1aLbOeuqzd
	SXTdRjqX+/8g4mYwEkAbvzvp1SILrj07iHhpEglYmEK6SiMCjf8Nq12129Ygot3rpI++ufrD03w
	dl3MPdSqU2anN0adHuNPvEZue27ciSzD57iBJMdjZvANMXVnZYlwOvhLvqT6zZJtnbVn0IF+XZr
	8piKpsijA9uOMx/HEubFpvKvgk27qHir+nxfX7+G39nHh98M9UyDlmJzuYRAbGwkrCAV4vS9LtA
	qenjHeKj22jnGhEF4o8tfPNsT/R+7MYKCQNvrcPLiHKIhtjNouFTiaD8RWBoh6UuMQqbIuuWtG9
	RLUh0yCraxkiZ40HCleNr6fdy7sKBKS8nCu2hC9qsbUgbExirUN8hQTAof+4zSnO+00XB5OhORc
	Ws4oviAvpN+xh5VSgAIuF0YZtOHaS4GmA9woCia/JKWovp67JX/b+tr2dJePhEv1uYZPmjSw==
X-Received: by 2002:a17:903:fb0:b0:2ca:e565:7b15 with SMTP id
 d9443c01a7336-2d015c91d62mr18424095ad.10.1785230301466;
        Tue, 28 Jul 2026 02:18:21 -0700 (PDT)
Received: from wenstp920.tpe.corp.google.com
 ([2a00:79e0:201d:8:c85f:b2bb:a077:d544])
        by smtp.gmail.com with ESMTPSA id
 d9443c01a7336-2cfde81dcc1sm48616965ad.84.2026.07.28.02.18.17
        (version=TLS1_3 cipher=TLS_AES_256_GCM_SHA384 bits=256/256);
        Tue, 28 Jul 2026 02:18:20 -0700 (PDT)
From: Chen-Yu Tsai <wenst@chromium.org>
To: Stephen Boyd <sboyd@kernel.org>,
	Brian Masney <bmasney@redhat.com>,
	Matthias Brugger <matthias.bgg@gmail.com>,
	AngeloGioacchino Del Regno <angelogioacchino.delregno@collabora.com>,
	Alessio Belle <alessio.belle@imgtec.com>,
	Luigi Santivetti <luigi.santivetti@imgtec.com>,
	Maarten Lankhorst <maarten.lankhorst@linux.intel.com>,
	Maxime Ripard <mripard@kernel.org>,
	Thomas Zimmermann <tzimmermann@suse.de>
Cc: Chen-Yu Tsai <wenst@chromium.org>,
	David Airlie <airlied@gmail.com>,
	Simona Vetter <simona@ffwll.ch>,
	linux-clk@vger.kernel.org,
	devicetree@vger.kernel.org,
	linux-mediatek@lists.infradead.org,
	linux-arm-kernel@lists.infradead.org,
	imagination@lists.freedesktop.org,
	dri-devel@lists.freedesktop.org,
	linux-kernel@vger.kernel.org,
	Icenowy Zheng <zhengxingda@iscas.ac.cn>,
	YoungJoon Lee <getfeus@gmail.com>,
	Conor Dooley <conor.dooley@microchip.com>
Subject: [PATCH v4 1/5] dt-bindings: clock: mediatek: Add mt8173 mfgtop
Date: Tue, 28 Jul 2026 17:17:59 +0800
Message-ID: <20260728091804.382753-2-wenst@chromium.org>
X-Mailer: git-send-email 2.55.0.229.g6434b31f56-goog
In-Reply-To: <20260728091804.382753-1-wenst@chromium.org>
References: <20260728091804.382753-1-wenst@chromium.org>
Precedence: bulk
X-Mailing-List: linux-clk@vger.kernel.org
List-Id: <linux-clk.vger.kernel.org>
List-Subscribe: <mailto:linux-clk+subscribe@vger.kernel.org>
List-Unsubscribe: <mailto:linux-clk+unsubscribe@vger.kernel.org>
MIME-Version: 1.0

The MFG (GPU) block on the MT8173 has a small glue layer, named MFG_TOP
in the datasheet, that contains clock gates, some power sequence signal
delays, and other unknown registers that get toggled when the GPU is
powered on.

The clock gates are exposed as clocks provided by a clock controller,
while the power sequencing bits are exposed as one singular power domain.

Reviewed-by: Conor Dooley <conor.dooley@microchip.com>
Reviewed-by: AngeloGioacchino Del Regno <angelogioacchino.delregno@collabora.com>
Signed-off-by: Chen-Yu Tsai <wenst@chromium.org>
---
 .../clock/mediatek,mt8173-mfgtop.yaml         | 70 +++++++++++++++++++
 include/dt-bindings/clock/mt8173-clk.h        |  7 ++
 2 files changed, 77 insertions(+)
 create mode 100644 Documentation/devicetree/bindings/clock/mediatek,mt8173-mfgtop.yaml

diff --git a/Documentation/devicetree/bindings/clock/mediatek,mt8173-mfgtop.yaml b/Documentation/devicetree/bindings/clock/mediatek,mt8173-mfgtop.yaml
new file mode 100644
index 000000000000..03db1ee9e594
--- /dev/null
+++ b/Documentation/devicetree/bindings/clock/mediatek,mt8173-mfgtop.yaml
@@ -0,0 +1,70 @@
+# SPDX-License-Identifier: (GPL-2.0 OR BSD-2-Clause)
+%YAML 1.2
+---
+
+$id: http://devicetree.org/schemas/clock/mediatek,mt8173-mfgtop.yaml#
+$schema: http://devicetree.org/meta-schemas/core.yaml#
+
+title: MediaTek MT8173 MFG TOP controller
+
+maintainers:
+  - AngeloGioacchino Del Regno <angelogioacchino.delregno@collabora.com>
+
+description:
+  The MFG TOP glue layer controls various signals going to the MFG (GPU)
+  block on the MT8173.
+
+properties:
+  compatible:
+    const: mediatek,mt8173-mfgtop
+
+  reg:
+    maxItems: 1
+
+  clocks:
+    maxItems: 4
+
+  clock-names:
+    items:
+      - const: sys
+      - const: mem
+      - const: core
+      - const: clk26m
+
+  power-domains:
+    maxItems: 1
+
+  '#clock-cells':
+    const: 1
+
+  '#power-domain-cells':
+    const: 0
+
+required:
+  - compatible
+  - reg
+  - clocks
+  - clock-names
+  - power-domains
+  - '#clock-cells'
+  - '#power-domain-cells'
+
+additionalProperties: false
+
+examples:
+  - |
+        #include <dt-bindings/clock/mt8173-clk.h>
+        #include <dt-bindings/power/mt8173-power.h>
+
+        clock-controller@13fff000 {
+            compatible = "mediatek,mt8173-mfgtop";
+            reg = <0x13fff000 0x1000>;
+            clocks = <&topckgen CLK_TOP_AXI_MFG_IN_SEL>,
+                     <&topckgen CLK_TOP_MEM_MFG_IN_SEL>,
+                     <&topckgen CLK_TOP_MFG_SEL>,
+                     <&clk26m>;
+            clock-names = "sys", "mem", "core", "clk26m";
+            power-domains = <&spm MT8173_POWER_DOMAIN_MFG>;
+            #clock-cells = <1>;
+            #power-domain-cells = <0>;
+        };
diff --git a/include/dt-bindings/clock/mt8173-clk.h b/include/dt-bindings/clock/mt8173-clk.h
index 3d00c98b9654..89e982f771db 100644
--- a/include/dt-bindings/clock/mt8173-clk.h
+++ b/include/dt-bindings/clock/mt8173-clk.h
@@ -243,6 +243,13 @@
 #define CLK_IMG_FD			7
 #define CLK_IMG_NR_CLK			8
 
+/* MFG_SYS */
+
+#define CLK_MFG_AXI			0
+#define CLK_MFG_MEM			1
+#define CLK_MFG_G3D			2
+#define CLK_MFG_26M			3
+
 /* MM_SYS */
 
 #define CLK_MM_SMI_COMMON		1

From patchwork Tue Jul 28 09:18:00 2026
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Chen-Yu Tsai <wenst@chromium.org>
X-Patchwork-Id: 14715042
Received: from mail-pl1-f172.google.com (mail-pl1-f172.google.com
 [209.85.214.172])
	(using TLSv1.2 with cipher ECDHE-RSA-AES128-GCM-SHA256 (128/128 bits))
	(No client certificate requested)
	by smtp.subspace.kernel.org (Postfix) with ESMTPS id EFBF4420476
	for <linux-clk@vger.kernel.org>; Tue, 28 Jul 2026 09:18:26 +0000 (UTC)
Authentication-Results: smtp.subspace.kernel.org;
 arc=none smtp.client-ip=209.85.214.172
ARC-Seal: i=1; a=rsa-sha256; d=subspace.kernel.org; s=arc-20240116;
	t=1785230309; cv=none;
 b=h+D3STVsi8u7/6PkDxb0jz/Weo8e2L4ImeXhbv/r7Ui7SrHeQfVoYnPYn03sigREWfzY9b6qBaCn66TYsK2wYaD51tQPAQy0pOzL1dY4WKnWXWRtWFPLNd0WdYTVcB6pXQgQBVB0lbo6iwPOlbOA+F6kbSYckfeRWZxFOs46SN8=
ARC-Message-Signature: i=1; a=rsa-sha256; d=subspace.kernel.org;
	s=arc-20240116; t=1785230309; c=relaxed/simple;
	bh=A146EED0+59ltr/PtWbjztumGFMIYQ/iLKAM1rCYKfA=;
	h=From:To:Cc:Subject:Date:Message-ID:In-Reply-To:References:
	 MIME-Version;
 b=mvsXIyemHYKiwRQlWv41J6eFK1d+AnFYKhrQgIn1KDZfw+NHIvioNd64bu585c6kbHybkYolQ0d4guk36ArdQS4ahwxy/lnI4wgSyrH87EdpzyUB0SIqhtSHugiVqb9v3APeNarjHDRh3iUBiqXKXFnbjEBWClhPOp1I5UCsGfA=
ARC-Authentication-Results: i=1; smtp.subspace.kernel.org;
 dmarc=pass (p=none dis=none) header.from=chromium.org;
 spf=pass smtp.mailfrom=chromium.org;
 dkim=pass (1024-bit key) header.d=chromium.org header.i=@chromium.org
 header.b=oCthohki; arc=none smtp.client-ip=209.85.214.172
Authentication-Results: smtp.subspace.kernel.org;
 dmarc=pass (p=none dis=none) header.from=chromium.org
Authentication-Results: smtp.subspace.kernel.org;
 spf=pass smtp.mailfrom=chromium.org
Authentication-Results: smtp.subspace.kernel.org;
	dkim=pass (1024-bit key) header.d=chromium.org header.i=@chromium.org
 header.b="oCthohki"
Received: by mail-pl1-f172.google.com with SMTP id
 d9443c01a7336-2cf52d15d88so36772315ad.2
        for <linux-clk@vger.kernel.org>; Tue, 28 Jul 2026 02:18:26 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=chromium.org; s=google; t=1785230306; x=1785835106;
 darn=vger.kernel.org;
        h=content-transfer-encoding:mime-version:references:in-reply-to
         :message-id:date:subject:cc:to:from:from:to:cc:subject:date
         :message-id:reply-to:content-type;
        bh=49h1VdeF+Zg75pYUFKdBO+lS3yePgIwfgX1PP2SoBhU=;
        b=oCthohkiYYSwM6Nts0fHBThpccNoQhg01pvfsNL8PpKYjGjyXz7Jj9k1/h6KVTcgua
         Mj1VTFY9I3U4HQIYafE3fl88zxh0gWfrUMojMo5gusNeBLiAMdoKq3f+6zH0Z2XtE3X8
         jY1bx+5d2TG81hH3vINL03H+DGXilTosu1inQ=
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20251104; t=1785230306; x=1785835106;
        h=content-transfer-encoding:mime-version:references:in-reply-to
         :message-id:date:subject:cc:to:from:x-gm-gg:x-gm-message-state:from
         :to:cc:subject:date:message-id:reply-to:content-type;
        bh=49h1VdeF+Zg75pYUFKdBO+lS3yePgIwfgX1PP2SoBhU=;
        b=O56EGYEf+UF72pFAUE/g8nKpVIgrj/n2G8HviFlrMKO86uR7eWQC+FWJ8fG0YTJ2jE
         70Y+01DaAHWCSE8IDGzm92KMbAEJT7GKCPtkbr2hfoX0/7yXrJ3kcfHIfQHvTVvkJV1X
         Jle5IRzD8HcDM9mRKEIWZ+bZDNanfNKpy6JB/yWj0HLS58NMb2E9xYQgdtJz+KqqmwVr
         LvRLr6CPeX8z6nlnYgI22IwJITk9HKSNFa6LDqNcGrck3VTx8Hk1aSKPlUidnqo4nJ2g
         cQY6J49XJLCDw+++V2xmcsoawo5ydRjvH2LGwYGRvML9c12/0iB7lnwj2upCOvr6rnQL
         TlMw==
X-Forwarded-Encrypted: i=1;
 AHgh+Rr9rAXndU/y0qjiIPt/VzXI8+GZf51Tg6gUKRBOTKhkAUmcR8EVQGlev2j36I4ykju4lAfdYleViEQ=@vger.kernel.org
X-Gm-Message-State: AOJu0YwEywr76sdImakzhNO+4BaZs0WNxZ9MI+YqqDjNKY9vsSWzuFDm
	RP1GEoliepj4PUE9W6KtGDdhQM6c595z/QMYhlXJn+34f1danE5t2amnh8qwNyRorg==
X-Gm-Gg: AR+sD11qv5YjKlXA0oAVo8JOH7y+9D7r11IbjHJG3s17kxOdf0ysRDELOQgOG2pxWP9
	fg1vA4NQBDPbOQG35HtsBKHggRTi8MolUBzRrLAzEjHaqxOYbFAW2PWGICRTvovuGitbFE51baD
	5KCskYdEhsLOM//8KrliNYokoKrLpbcIiyHaId2Swoj43V+u2wPySSXAztFpx6TWhK8dJ7Hx2T2
	ZN29+wup4rZ4GY67K57FHTIAIfpLOV46Sx3rT/uPi/xvvR4HZznbnSickJtHv/ZKN0HfF6gk+4j
	ba77NYRvZsUzVuneTLtSYi70bH/YcGrzQXy/mfEMZoKk0ENEygM9dPAzbqbRv7FVLDxcoM0Qr45
	XKNrcvmHT5kZPWD++WjNWOHZ760aafXGnnMUMbtQ9U09Lz79WRGkts+jdypV5S4H+3rzoStuNOr
	mS3HPZFXtlMtBLEIsZTRftuDC4ltNR5qG/45ictjqh+hNe+L+PgrEPzy8xiEg=
X-Received: by 2002:a17:903:2f0e:b0:2c9:c517:d070 with SMTP id
 d9443c01a7336-2d015ac45camr19804365ad.3.1785230306195;
        Tue, 28 Jul 2026 02:18:26 -0700 (PDT)
Received: from wenstp920.tpe.corp.google.com
 ([2a00:79e0:201d:8:c85f:b2bb:a077:d544])
        by smtp.gmail.com with ESMTPSA id
 d9443c01a7336-2cfde81dcc1sm48616965ad.84.2026.07.28.02.18.21
        (version=TLS1_3 cipher=TLS_AES_256_GCM_SHA384 bits=256/256);
        Tue, 28 Jul 2026 02:18:25 -0700 (PDT)
From: Chen-Yu Tsai <wenst@chromium.org>
To: Stephen Boyd <sboyd@kernel.org>,
	Brian Masney <bmasney@redhat.com>,
	Matthias Brugger <matthias.bgg@gmail.com>,
	AngeloGioacchino Del Regno <angelogioacchino.delregno@collabora.com>,
	Alessio Belle <alessio.belle@imgtec.com>,
	Luigi Santivetti <luigi.santivetti@imgtec.com>,
	Maarten Lankhorst <maarten.lankhorst@linux.intel.com>,
	Maxime Ripard <mripard@kernel.org>,
	Thomas Zimmermann <tzimmermann@suse.de>
Cc: Chen-Yu Tsai <wenst@chromium.org>,
	David Airlie <airlied@gmail.com>,
	Simona Vetter <simona@ffwll.ch>,
	linux-clk@vger.kernel.org,
	devicetree@vger.kernel.org,
	linux-mediatek@lists.infradead.org,
	linux-arm-kernel@lists.infradead.org,
	imagination@lists.freedesktop.org,
	dri-devel@lists.freedesktop.org,
	linux-kernel@vger.kernel.org,
	Icenowy Zheng <zhengxingda@iscas.ac.cn>,
	YoungJoon Lee <getfeus@gmail.com>
Subject: [PATCH v4 2/5] clk: mediatek: Add mt8173-mfgtop driver
Date: Tue, 28 Jul 2026 17:18:00 +0800
Message-ID: <20260728091804.382753-3-wenst@chromium.org>
X-Mailer: git-send-email 2.55.0.229.g6434b31f56-goog
In-Reply-To: <20260728091804.382753-1-wenst@chromium.org>
References: <20260728091804.382753-1-wenst@chromium.org>
Precedence: bulk
X-Mailing-List: linux-clk@vger.kernel.org
List-Id: <linux-clk.vger.kernel.org>
List-Subscribe: <mailto:linux-clk+subscribe@vger.kernel.org>
List-Unsubscribe: <mailto:linux-clk+unsubscribe@vger.kernel.org>
MIME-Version: 1.0

The MFG (GPU) block on the MT8173 has a small glue layer, named MFG_TOP
in the datasheet, that contains clock gates, some power sequence signal
delays, and other unknown registers that get toggled when the GPU is
powered on.

The clock gates are exposed as clocks provided by a clock controller,
while the power sequencing bits are exposed as one singular power domain.

Tested-by: Icenowy Zheng <zhengxingda@iscas.ac.cn>
Reviewed-by: Brian Masney <bmasney@redhat.com>
Reviewed-by: AngeloGioacchino Del Regno <angelogioacchino.delregno@collabora.com>
Signed-off-by: Chen-Yu Tsai <wenst@chromium.org>
---
Changes since v3:
- Changed pm_runtime_put() to pm_runtime_put_sync()
- Added missing of_node_put() in remove function

Changes since v2:
- Made COMMON_CLK_MT8173_MFGTOP depend on PM
  - Needed since the driver implements PM domains using the generic PM
    domain library, which also depends on PM
  - Fixes build breakage (kernel test robot)
- Fixed "RST_DELAY_CNT" name (Brian)
- Dropped unused mfg_desc (Brian)
- Added check of clk_prepare_enable()'s return value in
  clk_mt8173_mfgtop_power_on() (Brian)
- Saved error value for return in IS_ERR(data->clk_26m) branch
  (Dan Carpenter / kernel test robot w/ smatch)
---
 drivers/clk/mediatek/Kconfig             |  10 +
 drivers/clk/mediatek/Makefile            |   1 +
 drivers/clk/mediatek/clk-mt8173-mfgtop.c | 242 +++++++++++++++++++++++
 3 files changed, 253 insertions(+)
 create mode 100644 drivers/clk/mediatek/clk-mt8173-mfgtop.c

diff --git a/drivers/clk/mediatek/Kconfig b/drivers/clk/mediatek/Kconfig
index 2c09fd729bab..8eea3766a910 100644
--- a/drivers/clk/mediatek/Kconfig
+++ b/drivers/clk/mediatek/Kconfig
@@ -537,6 +537,16 @@ config COMMON_CLK_MT8173_IMGSYS
 	help
 	  This driver supports MediaTek MT8173 imgsys clocks.
 
+config COMMON_CLK_MT8173_MFGTOP
+	tristate "Clock and power driver for MediaTek MT8173 mfgtop"
+	depends on COMMON_CLK_MT8173
+	depends on PM
+	default COMMON_CLK_MT8173
+	select PM_GENERIC_DOMAINS
+	select PM_GENERIC_DOMAINS_OF
+	help
+	  This driver supports MediaTek MT8173 mfgtop clocks and power domain.
+
 config COMMON_CLK_MT8173_MMSYS
        tristate "Clock driver for MediaTek MT8173 mmsys"
        depends on COMMON_CLK_MT8173
diff --git a/drivers/clk/mediatek/Makefile b/drivers/clk/mediatek/Makefile
index d8736a060dbd..892a54eeb281 100644
--- a/drivers/clk/mediatek/Makefile
+++ b/drivers/clk/mediatek/Makefile
@@ -82,6 +82,7 @@ obj-$(CONFIG_COMMON_CLK_MT8167_VDECSYS) += clk-mt8167-vdec.o
 obj-$(CONFIG_COMMON_CLK_MT8173) += clk-mt8173-apmixedsys.o clk-mt8173-infracfg.o \
 				   clk-mt8173-pericfg.o clk-mt8173-topckgen.o
 obj-$(CONFIG_COMMON_CLK_MT8173_IMGSYS) += clk-mt8173-img.o
+obj-$(CONFIG_COMMON_CLK_MT8173_MFGTOP) += clk-mt8173-mfgtop.o
 obj-$(CONFIG_COMMON_CLK_MT8173_MMSYS) += clk-mt8173-mm.o
 obj-$(CONFIG_COMMON_CLK_MT8173_VDECSYS) += clk-mt8173-vdecsys.o
 obj-$(CONFIG_COMMON_CLK_MT8173_VENCSYS) += clk-mt8173-vencsys.o
diff --git a/drivers/clk/mediatek/clk-mt8173-mfgtop.c b/drivers/clk/mediatek/clk-mt8173-mfgtop.c
new file mode 100644
index 000000000000..5669ca9954dc
--- /dev/null
+++ b/drivers/clk/mediatek/clk-mt8173-mfgtop.c
@@ -0,0 +1,242 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * Copyright (c) 2024 Google LLC
+ * Author: Chen-Yu Tsai <wenst@chromium.org>
+ *
+ * Based on driver in downstream ChromeOS v5.15 kernel.
+ *
+ * Copyright (c) 2014 MediaTek Inc.
+ * Author: Chiawen Lee <chiawen.lee@mediatek.com>
+ */
+
+#include <dt-bindings/clock/mt8173-clk.h>
+
+#include <linux/bitfield.h>
+#include <linux/clk.h>
+#include <linux/mfd/syscon.h>
+#include <linux/module.h>
+#include <linux/of.h>
+#include <linux/platform_device.h>
+#include <linux/pm_domain.h>
+#include <linux/pm_runtime.h>
+#include <linux/regmap.h>
+
+#include "clk-gate.h"
+#include "clk-mtk.h"
+
+static const struct mtk_gate_regs mfg_cg_regs = {
+	.sta_ofs = 0x0000,
+	.clr_ofs = 0x0008,
+	.set_ofs = 0x0004,
+};
+
+#define GATE_MFG(_id, _name, _parent, _shift, _flags)	\
+	GATE_MTK_FLAGS(_id, _name, _parent, &mfg_cg_regs, _shift, &mtk_clk_gate_ops_setclr, _flags)
+
+/* TODO: The block actually has dividers for the core and mem clocks. */
+static const struct mtk_gate mfg_clks[] = {
+	GATE_MFG(CLK_MFG_AXI, "mfg_axi", "axi_mfg_in_sel", 0, CLK_SET_RATE_PARENT),
+	GATE_MFG(CLK_MFG_MEM, "mfg_mem", "mem_mfg_in_sel", 1, CLK_SET_RATE_PARENT),
+	GATE_MFG(CLK_MFG_G3D, "mfg_g3d", "mfg_sel", 2, CLK_SET_RATE_PARENT),
+	GATE_MFG(CLK_MFG_26M, "mfg_26m", "clk26m", 3, 0),
+};
+
+struct mt8173_mfgtop_data {
+	struct clk_hw_onecell_data *clk_data;
+	struct regmap *regmap;
+	struct generic_pm_domain genpd;
+	struct of_phandle_args parent_pd, child_pd;
+	struct clk *clk_26m;
+};
+
+/* Delay count in clock cycles */
+#define MFG_ACTIVE_POWER_CON0	0x24
+ #define RST_B_DELAY_CNT	GENMASK(7, 0)	/* pwr_rst_b de-assert delay during power-up */
+ #define CLK_EN_DELAY_CNT	GENMASK(15, 8)	/* CLK_DIS deassert delay during power-up */
+ #define CLK_DIS_DELAY_CNT	GENMASK(23, 16)	/* CLK_DIS assert delay during power-down */
+ #define FORCE_ABORT		BIT(30)		/* write 1 to force abort a power event */
+ #define ACTIVE_PWRCTL_EN	BIT(31)		/* enable ACTIVE_POWER */
+
+#define MFG_ACTIVE_POWER_CON1	0x28
+ #define PWR_ON_S_DELAY_CNT	GENMASK(7, 0)	/* pwr_on_s assert delay during power-up */
+ #define ISO_DELAY_CNT		GENMASK(15, 8)	/* ISO assert delay during power-down */
+ #define ISOOFF_DELAY_CNT	GENMASK(23, 16)	/* ISO de-assert delay during power-up */
+ #define RST_DELAY_CNT		GENMASK(31, 24) /* pwr_rsb_b assert delay during power-down */
+
+static int clk_mt8173_mfgtop_power_on(struct generic_pm_domain *domain)
+{
+	struct mt8173_mfgtop_data *data = container_of(domain, struct mt8173_mfgtop_data, genpd);
+	int ret;
+
+	/* drives internal power management */
+	ret = clk_prepare_enable(data->clk_26m);
+	if (ret)
+		return ret;
+
+	/* Power on/off delays for various signals */
+	regmap_write(data->regmap, MFG_ACTIVE_POWER_CON0,
+		     FIELD_PREP(RST_B_DELAY_CNT, 77) |
+		     FIELD_PREP(CLK_EN_DELAY_CNT, 61) |
+		     FIELD_PREP(CLK_DIS_DELAY_CNT, 60) |
+		     FIELD_PREP(ACTIVE_PWRCTL_EN, 0));
+	regmap_write(data->regmap, MFG_ACTIVE_POWER_CON1,
+		     FIELD_PREP(PWR_ON_S_DELAY_CNT, 11) |
+		     FIELD_PREP(ISO_DELAY_CNT, 68) |
+		     FIELD_PREP(ISOOFF_DELAY_CNT, 69) |
+		     FIELD_PREP(RST_DELAY_CNT, 77));
+
+	/* Magic numbers related to core switch sequence and delays */
+	regmap_write(data->regmap, 0xe0, 0x7a710184);
+	regmap_write(data->regmap, 0xe4, 0x835f6856);
+	regmap_write(data->regmap, 0xe8, 0x002b0234);
+	regmap_write(data->regmap, 0xec, 0x80000000);
+	regmap_write(data->regmap, 0xa0, 0x08000000);
+
+	return 0;
+}
+
+static int clk_mt8173_mfgtop_power_off(struct generic_pm_domain *domain)
+{
+	struct mt8173_mfgtop_data *data = container_of(domain, struct mt8173_mfgtop_data, genpd);
+
+	/* Magic numbers related to core switch sequence and delays */
+	regmap_write(data->regmap, 0xec, 0);
+
+	/* drives internal power management */
+	clk_disable_unprepare(data->clk_26m);
+
+	return 0;
+}
+
+static int clk_mt8173_mfgtop_probe(struct platform_device *pdev)
+{
+	struct device *dev = &pdev->dev;
+	struct device_node *node = dev->of_node;
+	struct mt8173_mfgtop_data *data;
+	int ret;
+
+	data = devm_kzalloc(dev, sizeof(*data), GFP_KERNEL);
+	if (!data)
+		return -ENOMEM;
+
+	platform_set_drvdata(pdev, data);
+
+	data->clk_data = mtk_devm_alloc_clk_data(dev, ARRAY_SIZE(mfg_clks));
+	if (!data->clk_data)
+		return -ENOMEM;
+
+	/* MTK clock gates also uses regmap */
+	data->regmap = device_node_to_regmap(node);
+	if (IS_ERR(data->regmap))
+		return dev_err_probe(dev, PTR_ERR(data->regmap), "Failed to get regmap\n");
+
+	data->child_pd.np = node;
+	data->child_pd.args_count = 0;
+	ret = of_parse_phandle_with_args(node, "power-domains", "#power-domain-cells", 0,
+					 &data->parent_pd);
+	if (ret)
+		return dev_err_probe(dev, ret, "Failed to parse power domain\n");
+
+	devm_pm_runtime_enable(dev);
+	/*
+	 * Do a pm_runtime_resume_and_get() to workaround a possible
+	 * deadlock between clk_register() and the genpd framework.
+	 */
+	ret = pm_runtime_resume_and_get(dev);
+	if (ret) {
+		dev_err_probe(dev, ret, "Failed to runtime resume device\n");
+		goto put_of_node;
+	}
+
+	ret = mtk_clk_register_gates(dev, node, mfg_clks, ARRAY_SIZE(mfg_clks),
+				     data->clk_data);
+	if (ret) {
+		dev_err_probe(dev, ret, "Failed to register clock gates\n");
+		goto put_pm_runtime;
+	}
+
+	data->clk_26m = clk_hw_get_clk(data->clk_data->hws[CLK_MFG_26M], "26m");
+	if (IS_ERR(data->clk_26m)) {
+		ret = dev_err_probe(dev, PTR_ERR(data->clk_26m), "Failed to get 26 MHz clock\n");
+		goto unregister_clks;
+	}
+
+	ret = of_clk_add_hw_provider(node, of_clk_hw_onecell_get, data->clk_data);
+	if (ret) {
+		dev_err_probe(dev, ret, "Failed to add clk OF provider\n");
+		goto put_26m_clk;
+	}
+
+	data->genpd.name = "mfg-top";
+	data->genpd.power_on = clk_mt8173_mfgtop_power_on;
+	data->genpd.power_off = clk_mt8173_mfgtop_power_off;
+	ret = pm_genpd_init(&data->genpd, NULL, true);
+	if (ret) {
+		dev_err_probe(dev, ret, "Failed to add power domain\n");
+		goto del_clk_provider;
+	}
+
+	ret = of_genpd_add_provider_simple(node, &data->genpd);
+	if (ret) {
+		dev_err_probe(dev, ret, "Failed to add power domain OF provider\n");
+		goto remove_pd;
+	}
+
+	ret = of_genpd_add_subdomain(&data->parent_pd, &data->child_pd);
+	if (ret) {
+		dev_err_probe(dev, ret, "Failed to link PM domains\n");
+		goto del_pd_provider;
+	}
+
+	pm_runtime_put(dev);
+	return 0;
+
+del_pd_provider:
+	of_genpd_del_provider(node);
+remove_pd:
+	pm_genpd_remove(&data->genpd);
+del_clk_provider:
+	of_clk_del_provider(node);
+put_26m_clk:
+	clk_put(data->clk_26m);
+unregister_clks:
+	mtk_clk_unregister_gates(mfg_clks, ARRAY_SIZE(mfg_clks), data->clk_data);
+put_pm_runtime:
+	pm_runtime_put_sync(dev);
+put_of_node:
+	of_node_put(data->parent_pd.np);
+	return ret;
+}
+
+static void clk_mt8173_mfgtop_remove(struct platform_device *pdev)
+{
+	struct mt8173_mfgtop_data *data = platform_get_drvdata(pdev);
+	struct device_node *node = pdev->dev.of_node;
+
+	of_genpd_remove_subdomain(&data->parent_pd, &data->child_pd);
+	of_genpd_del_provider(node);
+	pm_genpd_remove(&data->genpd);
+	of_clk_del_provider(node);
+	clk_put(data->clk_26m);
+	mtk_clk_unregister_gates(mfg_clks, ARRAY_SIZE(mfg_clks), data->clk_data);
+	of_node_put(data->parent_pd.np);
+}
+
+static const struct of_device_id of_match_clk_mt8173_mfgtop[] = {
+	{ .compatible = "mediatek,mt8173-mfgtop" },
+	{ /* sentinel */ }
+};
+MODULE_DEVICE_TABLE(of, of_match_clk_mt8173_mfgtop);
+
+static struct platform_driver clk_mt8173_mfgtop_drv = {
+	.probe = clk_mt8173_mfgtop_probe,
+	.remove = clk_mt8173_mfgtop_remove,
+	.driver = {
+		.name = "clk-mt8173-mfgtop",
+		.of_match_table = of_match_clk_mt8173_mfgtop,
+	},
+};
+module_platform_driver(clk_mt8173_mfgtop_drv);
+
+MODULE_DESCRIPTION("MediaTek MT8173 mfgtop clock driver");
+MODULE_LICENSE("GPL");

From patchwork Tue Jul 28 09:18:01 2026
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Chen-Yu Tsai <wenst@chromium.org>
X-Patchwork-Id: 14715043
Received: from mail-pl1-f182.google.com (mail-pl1-f182.google.com
 [209.85.214.182])
	(using TLSv1.2 with cipher ECDHE-RSA-AES128-GCM-SHA256 (128/128 bits))
	(No client certificate requested)
	by smtp.subspace.kernel.org (Postfix) with ESMTPS id 5057742047A
	for <linux-clk@vger.kernel.org>; Tue, 28 Jul 2026 09:18:31 +0000 (UTC)
Authentication-Results: smtp.subspace.kernel.org;
 arc=none smtp.client-ip=209.85.214.182
ARC-Seal: i=1; a=rsa-sha256; d=subspace.kernel.org; s=arc-20240116;
	t=1785230312; cv=none;
 b=n9bxT4fjjuM8cX8o0eqU23Nqs5FN2iLQsLgLEOXb9CsKfDflpuKVFDb/HkdEsKA3416NTspKxrYltQt34pBoBrUAq5teFh2Irr3wbdzqjE3KUtuqMLengwiwXFUh6/QAE6YWZDA/oK2uhCDGjGEBrotDlJZzNTcJkU6A9I/fB4A=
ARC-Message-Signature: i=1; a=rsa-sha256; d=subspace.kernel.org;
	s=arc-20240116; t=1785230312; c=relaxed/simple;
	bh=aIFMPuRdmJNDMfcNywwoN0iQ5141uX9JSYiOoIkP3dc=;
	h=From:To:Cc:Subject:Date:Message-ID:In-Reply-To:References:
	 MIME-Version;
 b=f1YRa4cqZqwZzC78oFFFwFcMtiRrKhF2W+yDZ7iKMBqrqI6+GTx8tn00MFfcrX/EWbUsJ0FG01pi7qrdyzBsm+9ZEGLLB2R4NICgGOlo6kTjEcvLM6DFyyoWjgM4C65FLQfeAvNDxSV3zgpfOS0HbjmyoxX6gNk9CZsYvm43On0=
ARC-Authentication-Results: i=1; smtp.subspace.kernel.org;
 dmarc=pass (p=none dis=none) header.from=chromium.org;
 spf=pass smtp.mailfrom=chromium.org;
 dkim=pass (1024-bit key) header.d=chromium.org header.i=@chromium.org
 header.b=fKDIGodD; arc=none smtp.client-ip=209.85.214.182
Authentication-Results: smtp.subspace.kernel.org;
 dmarc=pass (p=none dis=none) header.from=chromium.org
Authentication-Results: smtp.subspace.kernel.org;
 spf=pass smtp.mailfrom=chromium.org
Authentication-Results: smtp.subspace.kernel.org;
	dkim=pass (1024-bit key) header.d=chromium.org header.i=@chromium.org
 header.b="fKDIGodD"
Received: by mail-pl1-f182.google.com with SMTP id
 d9443c01a7336-2cce6a0c9c3so42099235ad.1
        for <linux-clk@vger.kernel.org>; Tue, 28 Jul 2026 02:18:31 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=chromium.org; s=google; t=1785230310; x=1785835110;
 darn=vger.kernel.org;
        h=content-transfer-encoding:mime-version:references:in-reply-to
         :message-id:date:subject:cc:to:from:from:to:cc:subject:date
         :message-id:reply-to:content-type;
        bh=OWdXeAE8uA9DV3aiiuBfjSr6VRPTM+UnFjc1T+sgsLA=;
        b=fKDIGodDmCiUWxnhOocxFne+yF+hl+36GD2rZZjgY8d7hvScNE0Ty06auHQjGtwoZI
         Pf9JLBh0vbfPVKQi/4qx0+Jb3t3zoxbWiUxM/gWZ3Yi1pAQbxosjPyWqD81AGlUeKVzm
         wuK6X2JzVlt0Cps6uDoKolfwGhlGq14OHJkTU=
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20251104; t=1785230310; x=1785835110;
        h=content-transfer-encoding:mime-version:references:in-reply-to
         :message-id:date:subject:cc:to:from:x-gm-gg:x-gm-message-state:from
         :to:cc:subject:date:message-id:reply-to:content-type;
        bh=OWdXeAE8uA9DV3aiiuBfjSr6VRPTM+UnFjc1T+sgsLA=;
        b=sLtDKlDifEK9bVE1VLZzbYzMW2LkCnHs1a5fsEgjILpdEdFVuWU4/MKqn28AA3hqDD
         ijU6Jh90F1+n0NoP8rzcCjpaMyr+jdvJzWkCURfL4BN4Z/GwSaknsm5qXUf8TmEevGTQ
         ENxyRZUBduzjpdrzsh2eIMaEkYyMQUJhjaFK1JHoQyXG/mTYQL8feDVVxnl2hgQOzsJ6
         9qSbNvKnda+rXs8KN1kZBZiGkIThLdDJ0Xv7wJblx2AAOAPu+wzhm/betiNxVszYjJN2
         KO8ONb9gDXbyQ9Q5obpJMVu/4Jidw9lzHMUy6W7PZO+hzQXC61HDnd1caSI1sizc1pfW
         3FDg==
X-Forwarded-Encrypted: i=1;
 AHgh+RrGTodtlg4bitTPAO6PqLenx6zFeyZVKiCUD1VE1gJGT6jvFKgvKiuyMByThv0WlJ14b03Txrv/9zA=@vger.kernel.org
X-Gm-Message-State: AOJu0YwZBfvozbmA0F6/LlpYd1gadmB48Gc6vDpMXs8K0bSpqUJ1vLiG
	tSHPZbghyZHDfhZ0LaNWB2HT25nYIy/5h3o/yWLnG0zEgbx436ngfTR4CvDcj8sygw==
X-Gm-Gg: AR+sD100qAmCr3dP6sAFddLs8N/D+PG6ZyY7+kWnFBn1//9A8REpC1KqJ5OuJVvQlSE
	VZN8Mwa3fOmsn3PkcgAmHBqF6Fx4azMhkjZBQd2EZ6pugeeFIvOzOkAPor7/4KMTTMTfhNEWiIv
	Jm1/RbVEDgjVczl6jsymssqR/sniJ+aDzI3G/l/7kvvRCvYWJvrBemEdzOoXgj5tgm51DsR4tTZ
	7kWuZpXssT6aV3nPAOtKL5j1wqgNkbvNBAHkA4h+sCBJmRcPjPL32gU9Lf6WcyHGWrpU/EKekz3
	BetJREebs7+Jr+qFn6UbVYIGET9vwXdUqny4M0RIvCC9ZRsO8c/4R1wVxCYa2ZLnwVAPMXAi1cS
	DaHQLrUrGmIwlzMCHlaSz+RlTlHmcinKRaqqPuYDuvWZKOQs7+DrtUEXkWhpW7BriR/LvgMVYba
	nC/QkOa+4Ni9LMnTbcO1ogZmm3QitS9Pe5+bUsBk/y6T3/roLrxQXDMkOEjy6+61fz5AAYfEZJa
	b2FbznQcAaM/IC/VBYbtFgK46MYrVkJFncGEq5FetSvvg==
X-Received: by 2002:a17:903:124a:b0:2c9:bf82:dd11 with SMTP id
 d9443c01a7336-2d015ace749mr18293065ad.7.1785230310643;
        Tue, 28 Jul 2026 02:18:30 -0700 (PDT)
Received: from wenstp920.tpe.corp.google.com
 ([2a00:79e0:201d:8:c85f:b2bb:a077:d544])
        by smtp.gmail.com with ESMTPSA id
 d9443c01a7336-2cfde81dcc1sm48616965ad.84.2026.07.28.02.18.26
        (version=TLS1_3 cipher=TLS_AES_256_GCM_SHA384 bits=256/256);
        Tue, 28 Jul 2026 02:18:30 -0700 (PDT)
From: Chen-Yu Tsai <wenst@chromium.org>
To: Stephen Boyd <sboyd@kernel.org>,
	Brian Masney <bmasney@redhat.com>,
	Matthias Brugger <matthias.bgg@gmail.com>,
	AngeloGioacchino Del Regno <angelogioacchino.delregno@collabora.com>,
	Alessio Belle <alessio.belle@imgtec.com>,
	Luigi Santivetti <luigi.santivetti@imgtec.com>,
	Maarten Lankhorst <maarten.lankhorst@linux.intel.com>,
	Maxime Ripard <mripard@kernel.org>,
	Thomas Zimmermann <tzimmermann@suse.de>
Cc: Chen-Yu Tsai <wenst@chromium.org>,
	David Airlie <airlied@gmail.com>,
	Simona Vetter <simona@ffwll.ch>,
	linux-clk@vger.kernel.org,
	devicetree@vger.kernel.org,
	linux-mediatek@lists.infradead.org,
	linux-arm-kernel@lists.infradead.org,
	imagination@lists.freedesktop.org,
	dri-devel@lists.freedesktop.org,
	linux-kernel@vger.kernel.org,
	Icenowy Zheng <zhengxingda@iscas.ac.cn>,
	YoungJoon Lee <getfeus@gmail.com>,
	Conor Dooley <conor.dooley@microchip.com>
Subject: [PATCH v4 3/5] dt-bindings: gpu: powervr-rogue: Add MediaTek MT8173
 GPU
Date: Tue, 28 Jul 2026 17:18:01 +0800
Message-ID: <20260728091804.382753-4-wenst@chromium.org>
X-Mailer: git-send-email 2.55.0.229.g6434b31f56-goog
In-Reply-To: <20260728091804.382753-1-wenst@chromium.org>
References: <20260728091804.382753-1-wenst@chromium.org>
Precedence: bulk
X-Mailing-List: linux-clk@vger.kernel.org
List-Id: <linux-clk.vger.kernel.org>
List-Subscribe: <mailto:linux-clk+subscribe@vger.kernel.org>
List-Unsubscribe: <mailto:linux-clk+unsubscribe@vger.kernel.org>
MIME-Version: 1.0

The MediaTek MT8173 comes with a PowerVR Rogue GX6250, which is one
of the Series6XT GPUs, another sub-family of the Rogue family.

This was part of the very first few versions of the PowerVR submission,
but was later dropped.

[1] https://lore.kernel.org/dri-devel/6eeccb26e09aad67fb30ffcd523c793a43c79c2a.camel@imgtec.com/

Reviewed-by: AngeloGioacchino Del Regno <angelogioacchino.delregno@collabora.com>
Reviewed-by: Conor Dooley <conor.dooley@microchip.com>
Signed-off-by: Chen-Yu Tsai <wenst@chromium.org>
---
 Documentation/devicetree/bindings/gpu/img,powervr-rogue.yaml | 1 +
 1 file changed, 1 insertion(+)

diff --git a/Documentation/devicetree/bindings/gpu/img,powervr-rogue.yaml b/Documentation/devicetree/bindings/gpu/img,powervr-rogue.yaml
index a6bccb600a57..e9218e0f363d 100644
--- a/Documentation/devicetree/bindings/gpu/img,powervr-rogue.yaml
+++ b/Documentation/devicetree/bindings/gpu/img,powervr-rogue.yaml
@@ -16,6 +16,7 @@ properties:
     oneOf:
       - items:
           - enum:
+              - mediatek,mt8173-gpu
               - renesas,r8a7796-gpu
               - renesas,r8a77961-gpu
           - const: img,img-gx6250

From patchwork Tue Jul 28 09:18:02 2026
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Chen-Yu Tsai <wenst@chromium.org>
X-Patchwork-Id: 14715044
Received: from mail-pl1-f178.google.com (mail-pl1-f178.google.com
 [209.85.214.178])
	(using TLSv1.2 with cipher ECDHE-RSA-AES128-GCM-SHA256 (128/128 bits))
	(No client certificate requested)
	by smtp.subspace.kernel.org (Postfix) with ESMTPS id CBE5341D203
	for <linux-clk@vger.kernel.org>; Tue, 28 Jul 2026 09:18:35 +0000 (UTC)
Authentication-Results: smtp.subspace.kernel.org;
 arc=none smtp.client-ip=209.85.214.178
ARC-Seal: i=1; a=rsa-sha256; d=subspace.kernel.org; s=arc-20240116;
	t=1785230317; cv=none;
 b=qzfwwS2GztwrC6qKkG51757eG2MOd216bTyA6aO8w76gbG0B0IYL0w+9ewWYS8VhxQ4Ndqv0/1VT8ffjBnLq6gcULrYYffEKupK7xGsgvahJcTk48aQNNtsBWymOBNw8kgxhR8aadLbQDBJ43ayI5So4lzi+V4H8vhn6XzvGs5k=
ARC-Message-Signature: i=1; a=rsa-sha256; d=subspace.kernel.org;
	s=arc-20240116; t=1785230317; c=relaxed/simple;
	bh=yhY7mLjZfGOCMMgNPHfCV2Pu5mgEkzZsyZuduXNnUDk=;
	h=From:To:Cc:Subject:Date:Message-ID:In-Reply-To:References:
	 MIME-Version;
 b=FT4Lf5IM7BhcffT6gzZsTglP3Kgk0bqLoQBY8jhv953QrgLxUb4OneiH0zT7WLGsGPYu2CK+UW/PV7pfxjNJxdhU5b6mUO+f8+xfPV4fdI9xmxdcuQm7gVl1+j3sgTq8KqVl4VpZHycHDUH8OvoCYF+lkA5KTTgjeipM0PiY3hU=
ARC-Authentication-Results: i=1; smtp.subspace.kernel.org;
 dmarc=pass (p=none dis=none) header.from=chromium.org;
 spf=pass smtp.mailfrom=chromium.org;
 dkim=pass (1024-bit key) header.d=chromium.org header.i=@chromium.org
 header.b=i7v0Ew4m; arc=none smtp.client-ip=209.85.214.178
Authentication-Results: smtp.subspace.kernel.org;
 dmarc=pass (p=none dis=none) header.from=chromium.org
Authentication-Results: smtp.subspace.kernel.org;
 spf=pass smtp.mailfrom=chromium.org
Authentication-Results: smtp.subspace.kernel.org;
	dkim=pass (1024-bit key) header.d=chromium.org header.i=@chromium.org
 header.b="i7v0Ew4m"
Received: by mail-pl1-f178.google.com with SMTP id
 d9443c01a7336-2d01663d816so3715435ad.1
        for <linux-clk@vger.kernel.org>; Tue, 28 Jul 2026 02:18:35 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=chromium.org; s=google; t=1785230315; x=1785835115;
 darn=vger.kernel.org;
        h=content-transfer-encoding:mime-version:references:in-reply-to
         :message-id:date:subject:cc:to:from:from:to:cc:subject:date
         :message-id:reply-to:content-type;
        bh=wxBpcLwQpOW2vKL5WhYL3BFfyuddH3BZqW5fz8lS/T8=;
        b=i7v0Ew4meJNgNnTbQUMhdPeZp2uWzz/mQ5qdVy7NtRNnYqn5qeHGjqAnkBX/ZkoQkQ
         VronlCISSWrZIOwCVUU4MItur+fynTB3iW5WcwG73alCnI+CbAIevvebO6DbSuorFjYo
         2vd+DVrghuoasJ1wnuLJzqrrXpX31v5ErgItk=
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20251104; t=1785230315; x=1785835115;
        h=content-transfer-encoding:mime-version:references:in-reply-to
         :message-id:date:subject:cc:to:from:x-gm-gg:x-gm-message-state:from
         :to:cc:subject:date:message-id:reply-to:content-type;
        bh=wxBpcLwQpOW2vKL5WhYL3BFfyuddH3BZqW5fz8lS/T8=;
        b=eYYAqPffqfcFx9owg3UQMmJQP825Fh2XmtcqwEub/QqZqGaf9NpQwaI5JDFeVXeiyB
         asy4wid2I0G5FymQ9OEJUoWtlB2ud5vUY8u8SjM2RQ3w5j6y2aJu1n/dbte/vlBvsxTh
         fwTGuUjc26nx6FG+TaR527z5hVsuVkqKDnNlayxp+XHZhJLs/p2HR7aPDFhpIVpfhGSJ
         MguhWYisdQQAVsDurgcgdS7ez4KfKi41ZQpVjK53n4El0sDhoi7NRSdDRZyax+sYvl8T
         IP1wML2Qp+tKaKcLHB59QOLjITgVo3BHmlTWILenBOVf9ji3zPiAKFLnN/TVvTdnJE8I
         tovw==
X-Forwarded-Encrypted: i=1;
 AHgh+RqPU6x3eXMnmQRWsUNjdRIfZWE41/aPAPC3yYmIqBEQZ6HOk0JjIOY31XGY5xgHmPSaFybqS9b9ULM=@vger.kernel.org
X-Gm-Message-State: AOJu0Ywn09rLOTKP2uEHg28c2Bgs2sdKSdPL3YbxOk+MsSXF/ZUWI8YW
	AwL50c2/TnHGZHpgLB7ZExH/pzQukYIxANbV4ObxK42IzIl9et13GheIiM7Jop1kVhVfkLqTOqP
	+PH9WwQ==
X-Gm-Gg: AR+sD11dAdY/OruMdQ2vUuqT09CqQPISiOjgGDKQbqWSaUKzdhptErepjL4ie1n0IKA
	46WUJOL1OfulFN0QZHODzGEyJJu2A/mUenOkXyBmtQ+adlwQS8rk6fCOzmUXFp+Dt8caA376VMa
	sdq26WAGjCC6oM7HJfF8bWRCw1U4Ez6HI+EFV0KwwLjn1Li1ppUphTWAW6/CkyLP1aIprtEP00l
	1VqXE2lexvfBrkjEXX3TrDxysRBulQYGVyu0Phyxl1r0P5Zy84NdUKn8JpV6joXmwQTzkI5rswx
	hYgyH+IWgVd8RCiF7aDTjTzOoeYqtwkFu5FlL/QIMkrGbyGUJjCLo5+PGiA4xTljZIzzrmHy36m
	XY5VAsFIQk7hkhJV3r1scdbDkjRRWc+BS8+NCHroObPKTRNf0WcLysgWliMvICXU0IBo7Df3KIy
	9SimZahDhSq4MmqEnlfnWuRqESbzMqyQxN4A4lbKzFzr/U8oqmmYLfNb0R6uo=
X-Received: by 2002:a17:903:8c6:b0:2cf:6f7f:55d8 with SMTP id
 d9443c01a7336-2d015cfb0e2mr18391495ad.37.1785230315142;
        Tue, 28 Jul 2026 02:18:35 -0700 (PDT)
Received: from wenstp920.tpe.corp.google.com
 ([2a00:79e0:201d:8:c85f:b2bb:a077:d544])
        by smtp.gmail.com with ESMTPSA id
 d9443c01a7336-2cfde81dcc1sm48616965ad.84.2026.07.28.02.18.30
        (version=TLS1_3 cipher=TLS_AES_256_GCM_SHA384 bits=256/256);
        Tue, 28 Jul 2026 02:18:34 -0700 (PDT)
From: Chen-Yu Tsai <wenst@chromium.org>
To: Stephen Boyd <sboyd@kernel.org>,
	Brian Masney <bmasney@redhat.com>,
	Matthias Brugger <matthias.bgg@gmail.com>,
	AngeloGioacchino Del Regno <angelogioacchino.delregno@collabora.com>,
	Alessio Belle <alessio.belle@imgtec.com>,
	Luigi Santivetti <luigi.santivetti@imgtec.com>,
	Maarten Lankhorst <maarten.lankhorst@linux.intel.com>,
	Maxime Ripard <mripard@kernel.org>,
	Thomas Zimmermann <tzimmermann@suse.de>
Cc: Chen-Yu Tsai <wenst@chromium.org>,
	David Airlie <airlied@gmail.com>,
	Simona Vetter <simona@ffwll.ch>,
	linux-clk@vger.kernel.org,
	devicetree@vger.kernel.org,
	linux-mediatek@lists.infradead.org,
	linux-arm-kernel@lists.infradead.org,
	imagination@lists.freedesktop.org,
	dri-devel@lists.freedesktop.org,
	linux-kernel@vger.kernel.org,
	Icenowy Zheng <zhengxingda@iscas.ac.cn>,
	YoungJoon Lee <getfeus@gmail.com>
Subject: [PATCH v4 4/5] arm64: dts: mediatek: mt8173: Fix MFG_ASYNC power
 domain clock
Date: Tue, 28 Jul 2026 17:18:02 +0800
Message-ID: <20260728091804.382753-5-wenst@chromium.org>
X-Mailer: git-send-email 2.55.0.229.g6434b31f56-goog
In-Reply-To: <20260728091804.382753-1-wenst@chromium.org>
References: <20260728091804.382753-1-wenst@chromium.org>
Precedence: bulk
X-Mailing-List: linux-clk@vger.kernel.org
List-Id: <linux-clk.vger.kernel.org>
List-Subscribe: <mailto:linux-clk+subscribe@vger.kernel.org>
List-Unsubscribe: <mailto:linux-clk+unsubscribe@vger.kernel.org>
MIME-Version: 1.0

The MFG_ASYNC domain, which is likely associated to the whole MFG block,
currently specifies clk26m as its domain clock. This is bogus, since the
clock is an external crystal with no controls. Also, the MFG block has
a independent CLK_TOP_AXI_MFG_IN_SEL clock, which according to the block
diagram, gates access to the hardware registers. Having this one as the
domain clock makes much more sense. This also fixes access to the MFGTOP
registers.

Change the MFG_ASYNC domain clock to CLK_TOP_AXI_MFG_IN_SEL.

Fixes: 8b6562644df9 ("arm64: dts: mediatek: Add mt8173 power domain controller")
Reviewed-by: AngeloGioacchino Del Regno <angelogioacchino.delregno@collabora.com>
Tested-by: Icenowy Zheng <zhengxingda@iscas.ac.cn>
Signed-off-by: Chen-Yu Tsai <wenst@chromium.org>
---
 arch/arm64/boot/dts/mediatek/mt8173.dtsi | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/arch/arm64/boot/dts/mediatek/mt8173.dtsi b/arch/arm64/boot/dts/mediatek/mt8173.dtsi
index 78c2ccd5be13..556e806b7787 100644
--- a/arch/arm64/boot/dts/mediatek/mt8173.dtsi
+++ b/arch/arm64/boot/dts/mediatek/mt8173.dtsi
@@ -505,7 +505,7 @@ power-domain@MT8173_POWER_DOMAIN_USB {
 				};
 				mfg_async: power-domain@MT8173_POWER_DOMAIN_MFG_ASYNC {
 					reg = <MT8173_POWER_DOMAIN_MFG_ASYNC>;
-					clocks = <&clk26m>;
+					clocks = <&topckgen CLK_TOP_AXI_MFG_IN_SEL>;
 					clock-names = "mfg";
 					#address-cells = <1>;
 					#size-cells = <0>;

From patchwork Tue Jul 28 09:18:03 2026
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Chen-Yu Tsai <wenst@chromium.org>
X-Patchwork-Id: 14715045
Received: from mail-pl1-f180.google.com (mail-pl1-f180.google.com
 [209.85.214.180])
	(using TLSv1.2 with cipher ECDHE-RSA-AES128-GCM-SHA256 (128/128 bits))
	(No client certificate requested)
	by smtp.subspace.kernel.org (Postfix) with ESMTPS id 2025541DDFC
	for <linux-clk@vger.kernel.org>; Tue, 28 Jul 2026 09:18:39 +0000 (UTC)
Authentication-Results: smtp.subspace.kernel.org;
 arc=none smtp.client-ip=209.85.214.180
ARC-Seal: i=1; a=rsa-sha256; d=subspace.kernel.org; s=arc-20240116;
	t=1785230321; cv=none;
 b=Rv7uDmZ31rq2tVrEMHuv6ZPaQ+CmE6Q9/8uZqnsHEe277y+Pq2PrNa742JVgsWyV+B9B8nnw+7lI/EVioB3AZw5BThr2JRaBgii0U36+MHCEmfAs4p66OVEEJfCjyGGLUWZqntRvuAf6UzFUiAaGOs5LZuM1qHLsMsqiJ0Yqlwg=
ARC-Message-Signature: i=1; a=rsa-sha256; d=subspace.kernel.org;
	s=arc-20240116; t=1785230321; c=relaxed/simple;
	bh=9Pf42jTYd4BOZWSzwEb2fi7Fbo1AsB7wRyp4TwTJs3s=;
	h=From:To:Cc:Subject:Date:Message-ID:In-Reply-To:References:
	 MIME-Version;
 b=s5fa+Yj0RTeH+1DdoXxJX1dZ6WoGsDc4yUhZKYkw5HlL7hAsSx6nyphvQBH/PMtv5NyyZ7gN3g0hzOSFj7fvFoRf1DjcZVpyGfkOAtYwOAwKa3eaZbtyCkvCVr4th3n7mgIvaTG8rKwdPgAqH8x0WjdKKxzygBD0/REzNZWWLfA=
ARC-Authentication-Results: i=1; smtp.subspace.kernel.org;
 dmarc=pass (p=none dis=none) header.from=chromium.org;
 spf=pass smtp.mailfrom=chromium.org;
 dkim=pass (1024-bit key) header.d=chromium.org header.i=@chromium.org
 header.b=XaPNyYAg; arc=none smtp.client-ip=209.85.214.180
Authentication-Results: smtp.subspace.kernel.org;
 dmarc=pass (p=none dis=none) header.from=chromium.org
Authentication-Results: smtp.subspace.kernel.org;
 spf=pass smtp.mailfrom=chromium.org
Authentication-Results: smtp.subspace.kernel.org;
	dkim=pass (1024-bit key) header.d=chromium.org header.i=@chromium.org
 header.b="XaPNyYAg"
Received: by mail-pl1-f180.google.com with SMTP id
 d9443c01a7336-2cf6d65d8a7so53122325ad.0
        for <linux-clk@vger.kernel.org>; Tue, 28 Jul 2026 02:18:39 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=chromium.org; s=google; t=1785230319; x=1785835119;
 darn=vger.kernel.org;
        h=content-transfer-encoding:mime-version:references:in-reply-to
         :message-id:date:subject:cc:to:from:from:to:cc:subject:date
         :message-id:reply-to:content-type;
        bh=AETGy+ANMs/4mEVJOXJB8nyO+6HDkmoHxo4KA1OMiIw=;
        b=XaPNyYAgYYfBJHWuxPo2imNXxmZHKGpNIz6moCpNat+oxEUJyfnX2EenK1zkW0t4LL
         RQlM/PgQKj+SEO7RGO+hocYUhX8KDgoQmNDrN6+7VGec2p2cFKOJ+Vl0cSYu0gFVOIx9
         3k/E/2CfptBus0BgkM0M9YNBkP6+ppqmImNV4=
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20251104; t=1785230319; x=1785835119;
        h=content-transfer-encoding:mime-version:references:in-reply-to
         :message-id:date:subject:cc:to:from:x-gm-gg:x-gm-message-state:from
         :to:cc:subject:date:message-id:reply-to:content-type;
        bh=AETGy+ANMs/4mEVJOXJB8nyO+6HDkmoHxo4KA1OMiIw=;
        b=mQKHEaNOZVYjk7cf2Gf4iJz3TbvgagSJqZcLZdUJyt4/xCXlO1qAkIA0cSn/gBmdrc
         JEtlxPJeqP/PD7Zyb8ZVOrdeGdVad01Ck07UgFY8HKMEWCJMPxtPSkiYD6DtHHUyZeZs
         d3QALxKruNIi6Mzd4poEcpNq5YcUQ+o9YYTi4vxlR5SWJhnAxhiXglTXgMaR5V7TEzWe
         +mwDHkh7McnCE8hf+4ZZ2gT7jZuKRi/ArNHu2+V4bvjIu3GAQ60ulADAVPD8sAbOA6kY
         OQUSdQj1kYPKan6c7/l+bFI/kZQ+W38syjJlWL8jLjKAeT9q2RfYst9TA7j3cu68pOtL
         Xubg==
X-Forwarded-Encrypted: i=1;
 AHgh+RoSWzWpOhMyHFRkzqXzQWWMN4s7YcnWjAlHdOgnZbEN0Jb/ktBpX323CggFbNqz5h77bluM9qe3+p8=@vger.kernel.org
X-Gm-Message-State: AOJu0YxJc6xifhYyeoTtwdg8uIXMW3i0fkm7C/msgifyzPfO19paIZdJ
	wiIAAfMQWrxDNGTfA4J87YnVlkfy+147/zXeGM5N+SItE1Inw1a94n31wcFE1ACGqA==
X-Gm-Gg: AR+sD11VzER6HPLXMIcW0pi5Ptc1+x2i4IfU4tcA30YyFAuLz7gjzvTizGXQf92Cp4u
	M5bHHjxS5b4br5BTYuLfB1GkY9n/mSFiADUw+4TpafaFNhZI6PulA+Hnq1WsFaEPyXwq7smbIQB
	maFwyWBrmY7zrb9dD5G80IcLaQgJFMBHK28cRFpMqEEtsrVU5pGsIssrEPG211lLn3F8Utgm/YW
	IOKBo8g3wyGyZ9+UlcIa0UB1LF5/vUrHnXop4IjFq1oCA++L9bPGwfMNxbmiNluh9f1mctCHHor
	28NG7rLu5BMidCrd+zoRekRtVZ3dNsZlXtFI+iZGXVKap4AhddQcH/5CdX36j+tO/9+PF7IN47v
	wZUblYYCuLwlCysMDTU8y0XWKOxNimUphL/I77Nzw1s0CRPr77fBPEBOW4Pdi+gqr60lV+r/PPM
	m6PshQcKrPAvA1ga+uk/zdsGaj7HPS3KfqbPd/kv3EQ7mS84i6orYV9dnkoKq+M9bCf5Fl1Q==
X-Received: by 2002:a17:903:19f0:b0:2c9:fc14:588f with SMTP id
 d9443c01a7336-2d015f024b4mr18169045ad.29.1785230319419;
        Tue, 28 Jul 2026 02:18:39 -0700 (PDT)
Received: from wenstp920.tpe.corp.google.com
 ([2a00:79e0:201d:8:c85f:b2bb:a077:d544])
        by smtp.gmail.com with ESMTPSA id
 d9443c01a7336-2cfde81dcc1sm48616965ad.84.2026.07.28.02.18.35
        (version=TLS1_3 cipher=TLS_AES_256_GCM_SHA384 bits=256/256);
        Tue, 28 Jul 2026 02:18:38 -0700 (PDT)
From: Chen-Yu Tsai <wenst@chromium.org>
To: Stephen Boyd <sboyd@kernel.org>,
	Brian Masney <bmasney@redhat.com>,
	Matthias Brugger <matthias.bgg@gmail.com>,
	AngeloGioacchino Del Regno <angelogioacchino.delregno@collabora.com>,
	Alessio Belle <alessio.belle@imgtec.com>,
	Luigi Santivetti <luigi.santivetti@imgtec.com>,
	Maarten Lankhorst <maarten.lankhorst@linux.intel.com>,
	Maxime Ripard <mripard@kernel.org>,
	Thomas Zimmermann <tzimmermann@suse.de>
Cc: Chen-Yu Tsai <wenst@chromium.org>,
	David Airlie <airlied@gmail.com>,
	Simona Vetter <simona@ffwll.ch>,
	linux-clk@vger.kernel.org,
	devicetree@vger.kernel.org,
	linux-mediatek@lists.infradead.org,
	linux-arm-kernel@lists.infradead.org,
	imagination@lists.freedesktop.org,
	dri-devel@lists.freedesktop.org,
	linux-kernel@vger.kernel.org,
	Icenowy Zheng <zhengxingda@iscas.ac.cn>,
	YoungJoon Lee <getfeus@gmail.com>
Subject: [PATCH v4 5/5] arm64: dts: mediatek: mt8173: Add GPU device nodes
Date: Tue, 28 Jul 2026 17:18:03 +0800
Message-ID: <20260728091804.382753-6-wenst@chromium.org>
X-Mailer: git-send-email 2.55.0.229.g6434b31f56-goog
In-Reply-To: <20260728091804.382753-1-wenst@chromium.org>
References: <20260728091804.382753-1-wenst@chromium.org>
Precedence: bulk
X-Mailing-List: linux-clk@vger.kernel.org
List-Id: <linux-clk.vger.kernel.org>
List-Subscribe: <mailto:linux-clk+subscribe@vger.kernel.org>
List-Unsubscribe: <mailto:linux-clk+unsubscribe@vger.kernel.org>
MIME-Version: 1.0

The MediaTek MT8173 comes with a PowerVR Rogue GX6250, which is part
of the Series6XT, another variation of the Rogue family of GPUs.

On top of the GPU is a glue layer that handles some clock and power
signals.

Add device nodes for both. While there is supposed to be two different
power domains to the GPU, MediaTek does not have any information on how
this is integrated internally. Assign the same GPU power domain to both.

Reviewed-by: AngeloGioacchino Del Regno <angelogioacchino.delregno@collabora.com>
Tested-by: Icenowy Zheng <zhengxingda@iscas.ac.cn>
Signed-off-by: Chen-Yu Tsai <wenst@chromium.org>
---
 arch/arm64/boot/dts/mediatek/mt8173.dtsi | 31 ++++++++++++++++++++++++
 1 file changed, 31 insertions(+)

diff --git a/arch/arm64/boot/dts/mediatek/mt8173.dtsi b/arch/arm64/boot/dts/mediatek/mt8173.dtsi
index 556e806b7787..169ba4fd183e 100644
--- a/arch/arm64/boot/dts/mediatek/mt8173.dtsi
+++ b/arch/arm64/boot/dts/mediatek/mt8173.dtsi
@@ -1002,6 +1002,37 @@ u2port1: usb-phy@11291000 {
 			};
 		};
 
+		gpu: gpu@13000000 {
+			compatible = "mediatek,mt8173-gpu", "img,img-gx6250",
+				     "img,img-rogue";
+			reg = <0 0x13000000 0 0x10000>;
+			interrupts = <GIC_SPI 217 IRQ_TYPE_LEVEL_LOW>;
+			clocks = <&mfgtop CLK_MFG_G3D>,
+				 <&mfgtop CLK_MFG_MEM>,
+				 <&mfgtop CLK_MFG_AXI>;
+			clock-names = "core", "mem", "sys";
+			/*
+			 * Binding and GPU core HW have two power domain
+			 * inputs, but MediaTek does not have information
+			 * on how this is done internally.
+			 */
+			power-domains = <&mfgtop>, <&mfgtop>;
+			power-domain-names = "a", "b";
+		};
+
+		mfgtop: clock-controller@13fff000 {
+			compatible = "mediatek,mt8173-mfgtop";
+			reg = <0 0x13fff000 0 0x1000>;
+			clocks = <&topckgen CLK_TOP_AXI_MFG_IN_SEL>,
+				 <&topckgen CLK_TOP_MEM_MFG_IN_SEL>,
+				 <&topckgen CLK_TOP_MFG_SEL>,
+				 <&clk26m>;
+			clock-names = "sys", "mem", "core", "clk26m";
+			power-domains = <&spm MT8173_POWER_DOMAIN_MFG>;
+			#clock-cells = <1>;
+			#power-domain-cells = <0>;
+		};
+
 		mmsys: syscon@14000000 {
 			compatible = "mediatek,mt8173-mmsys", "syscon";
 			reg = <0 0x14000000 0 0x1000>;
